### PR TITLE
HtmlEditor - add ImageUpload advanced functionality

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

File: js/ui/html_editor/modules/imageCursor.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+const clickEvent = addNamespace('dxclick', MODULE_NAMESPACE);
+
+let ImageCursorModule = BaseModule;
+
+if(Quill) {
+    ImageCursorModule = class ImageCursorModule extends BaseModule {
+        constructor(quill, options) {
+            super(quill, options);
+
+            this.quill = quill;
```
Comment (DokaRus) [inline js/ui/html_editor/modules/imageCursor.js]: There is no need to save quill instance. This is already done in the `devextreme-quill/core/module` - 
https://github.com/DevExpress/devextreme-quill/blob/4239db63edcd7daa5dfe2f0bb3ef4847e9cd8789/core/module.js#L3

File: js/ui/html_editor/modules/imageUpload.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+let ImageUploadModule = BaseModule;
+
+if(Quill) {
+    ImageUploadModule = class ImageUploadModule extends BaseModule {
+        constructor(quill, options) {
+            super(quill, options);
+
+            this.options = options;
+            this.quill = quill;
```
Comment (DokaRus) [inline js/ui/html_editor/modules/imageUpload.js]: There is no need to save quill instance and initial options. This is already done in the `devextreme-quill/core/module` - 
https://github.com/DevExpress/devextreme-quill/blob/4239db63edcd7daa5dfe2f0bb3ef4847e9cd8789/core/module.js#L3

File: js/ui/html_editor/modules/imageUpload.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        }
+
+        _initFileUploader() {
+            const $container = $('<div>')
+                .addClass(HIDDEN_FILE_UPLOADER_CLASS)
+                .appendTo(this._quillContainer);
+
+            const fileUploaderOptions = extend({}, getFileUploaderBaseOptions(), {
+                uploadUrl: this.options.uploadUrl,
+                onUploaded: this._onUploaded.bind(this),
```
Comment (DokaRus) [inline js/ui/html_editor/modules/imageUpload.js:64]: We should add a note to the documentation, which will warn against changing this option

File: js/ui/html_editor/modules/imageUpload.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            }, this.options.fileUploaderOptions);
+
+            this._fileUploader = this.editorInstance._createComponent($container, FileUploader, fileUploaderOptions);
+
+            return $container;
+        }
+
+        _onUploaded(data) {
+            const selection = this.quill.getSelection();
+            const pasteIndex = selection ? selection.index : this.quill.getLength();
```
Comment (DokaRus) [inline js/ui/html_editor/modules/imageUpload.js]:
```suggestion
            const { index: pasteIndex } = this.quill.getSelection() ?? { index: this.quill.getLength() }; 
```

File: js/ui/html_editor/modules/imageUpload.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                this.handleOptionChangeValue(value);
+                return;
+            }
+
+            if(option === 'fileUploadMode') {
+                this.options.fileUploadMode = value;
+                this._handleServerUpload();
+            } else if(option === 'fileUploaderOptions') {
+                this._fileUploader.option(value);
+            }
```
Comment (DokaRus) [inline js/ui/html_editor/modules/imageUpload.js]: Use may use `switch` for this

File: js/ui/html_editor/ui.html_editor.js [hunk 6]
```diff
@@ -434,15 +436,24 @@ const HtmlEditor = Editor.inherit({
         return this._$htmlContainer;
     },
 
+    _parseOptionsData(args) {
```
Comment (DokaRus) [inline js/ui/html_editor/ui.html_editor.js]:
```suggestion
    _prepareModuleOptions(args) {
```

File: js/ui/html_editor/modules/imageCursor.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                this._adjustSelection(e);
+            }
+        }
+
+        _isAllowedTarget(targetElement) {
+            return this._isImage(targetElement);
+        }
+
+        _isImage(targetElement) {
+            return targetElement.tagName.toUpperCase() === 'IMG';
```
Comment (Zedwag) [inline js/ui/html_editor/modules/imageCursor.js:39]: Isn't `tagName` property already always in upper case? https://developer.mozilla.org/en-US/docs/Web/API/Element/tagName

Comment (Zedwag) [reply js/ui/html_editor/modules/imageCursor.js:39]: Or it's for XHTML, all right then, sorry :)

File: js/ui/html_editor/modules/imageUpload.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            this._handleInsertImages(e, 'clipboardData');
+        }
+
+        _handleInsertImages(e, filesField) {
+            this.saveValueChangeEvent(e);
+            const files = Array.from(e.originalEvent[filesField].files || []);
+
+            const uploads = files;
+
+            if(uploads.length > 0) {
```
Comment (Zedwag) [inline js/ui/html_editor/modules/imageUpload.js]:
```suggestion
            if(uploads.length) {
```

**Review by Zedwag — DISMISSED** [review]

**Review by Zedwag — APPROVED** [review]

**Pull request merged**